fix: made the RECAPTCHA_SITE_KEY mandatory if  RECAPTCHA_SITE_KEY is …

Doraemon012 opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.

-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**

<!-- E.g. a bugfix, feature, refactoring, etc… -->
This PR is a Bugfix.

**Issue Number:**

Fixes #1625 

**Did you add tests for your changes?**

<!--Yes or No. Note: Add unit tests or automation tests for your code.-->
No

**Snapshots/Videos:**

1. For `RECAPCHA KEYS`:

https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/d965b5db-7770-4018-8639-70fcf1da8a07


[not_setting_recaptcha.webm](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/88614057-2a2b-426c-8385-ab4e697a9df2)

2. For `Congratulations message`:

[Conguratilations_message_no_error.webm](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/2f8dd0f7-eae0-45ed-91bf-022168a7847a)

[congrutilations_message_error.webm](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/6a299202-6676-45dc-9fc2-fc5b892f4d74)

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**
No

<!--Add link to Talawa-Docs.-->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

This pull request addresses the following issues:
1. **RECAPTCHA_SITE_KEY Mandatory Requirement**: Enforces the mandatory nature of `RECAPTCHA_SITE_KEY` when `RECAPTCHA_SECRET_KEY` is provided.
This ensures that users are prompted to set up a `RECAPTCHA_SITE_KEY` when `RECAPTCHA_SECRET_KEY` is set up.
2. **Congratulations Message Duplication**: Resolves the occasional duplication of the Congratulations message.

**Does this PR introduce a breaking change?**
This PR dose not produce any breaking change.
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

<!--Yes or No-->
Yes

palisadoes commented on setup.ts:
```diff
@@ -247,15 +247,11 @@ async function recaptcha(): Promise<void> {
-  console.log(
-    '\nSelect reCAPTCHA v2 and the "I`m not a robot" checkbox option'
-  );
-  console.log(
```
Remove. I ran the script based on this PR and the text still remains somewhere

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/923cedb7-2c33-4097-b43a-98486fd652ae)

palisadoes commented on setup.ts:
```diff
@@ -247,15 +247,11 @@ async function recaptcha(): Promise<void> {
+    console.log(
+      ` \nreCAPTCHA secret key already exists with the value ${process.env.RECAPTCHA_SECRET_KEY}`
+    );
+  }
```
If you enter a new secret, the old one still gets echoed to the screen

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/541d6adb-80a5-491c-b0f0-e1d7fa95a0f4)

palisadoes commented on setup.ts:
```diff
@@ -247,15 +247,11 @@ async function recaptcha(): Promise<void> {
-  console.log(
-    '\nAdd "localhost" in domains and accept the terms, then press submit'
-  );
+  if (process.env.RECAPTCHA_SECRET_KEY) {
```
If you enter the site key and it's accepted the value gets immediately updated in the `.env` file. If you make a mistake and hit <CTRL-C> then your mistake is recorded in the config. There must be a way to confirm all the values before committing them to the configuration file

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/2615c49f-1946-4a24-824d-fb7267806a94)

palisadoes commented on setup.ts:
```diff
@@ -418,7 +409,7 @@ async function main(): Promise<void> {
   if (process.env.ACCESS_TOKEN_SECRET) {
     console.log(
-      `\nAccess token secret already exists with the value:\n${process.env.ACCESS_TOKEN_SECRET}`
+      `\nAccess token secret already exists with the value: \n${process.env.ACCESS_TOKEN_SECRET}`
```
In all cases for all values, if the value already exists, then the default action for creating a new value must be "No", not "Yes"

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/d1bd4aa4-a560-4a43-97ab-7413f9280169)

palisadoes reviewed (changes requested):
See comments

Doraemon012 commented:
@palisadoes 
I've implemented the requested changes.
I've also included the fix of issue #1590.
Kindly review the updates.

Videos demonstrating the output:

1. For `reCAPTCHA keys`:

[recaptcha.webm](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/0fdc5cc3-dc54-4a68-ac57-3e037e22fb58)

3. For `.env` file:
    - Empty `.env` file:

    [empty_env.webm](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/d2a5e3e2-b9d9-46d7-8a16-efa3b3b2509a)

    - Deleted `.env` file:
    
    [deleted_env.webm](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/0a97b836-9648-4535-90ab-675cea6cd7f6)

    - Handling comments and deleted parameters
    
    
    [commented_and_deleted_parameters.webm](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/99781a3e-9047-42b5-91d0-d85d4d94b661)

palisadoes commented on setup.ts:
```diff
@@ -108,10 +120,7 @@ async function redisConfiguration(): Promise<void> {
 
     const config = dotenv.parse(fs.readFileSync(".env"));
     config.REDIS_URL = url;
```
1. I commented out the Redis URL and then ran the script
1. When prompted I used localhost for the hostname
1. I then used port 1235 as the port which was incorrect
1. The script said that Redis was functional
1. The script completed with the congratulations message
1. The script did not update the Redis URL

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/c2613bdb-95a1-4f4d-961d-56655baf2033)

palisadoes replied:
The Redis check is not functional. It detects Redis running on google.com

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/9c072a2f-24f6-44d1-99d5-3ce30dc58c60)

palisadoes reviewed (changes requested):
1. The new defaults when values are present are working well.
2. There needs to be some updates as described

palisadoes commented:
Please also add a prompt for the environment, development vs production. Will probably start alpha testing later this year

palisadoes commented:
@Doraemon012 

1. Any updates on your progress on this? 
2. It's a really important feature for us to implement

palisadoes reviewed (changes requested):
Also, there is some confusion with using the `LAST_RESORT_SUPERADMIN_EMAIL` parameter. Contributors don’t realize that they must register with this email address before they can log in.

Please make an adjustment that whenever the `LAST_RESORT_SUPERADMIN_EMAIL` parameter is modified, including when it is first used, that a message is given clearly letting them know that they must register with this email address before they can log in.

palisadoes commented:
Please fix the failing test

Doraemon012 commented:
> @Doraemon012
> 
> 1. Any updates on your progress on this?
> 2. It's a really important feature for us to implement

@palidoses i’m working on it, will complete it as soon as possible

Doraemon012 commented:
@palisadoes 
I've added a console log for `redis` which says that `Redis URL already exists with value: redis://password@host:port`.
There's no `REDIS_URL` field in the `.env` file.

Due to this one of the CodeQL test is failing.

Doraemon012 commented:
Here's the final output :
[output.webm](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/3ffcf0eb-f277-4a3e-a1f0-b255fbf0fdc4)

Doraemon012 commented:
> What can be done to fix? Why didn't it fail for the other prompts?
> 
> Please remember the registration message when configuring the email of last resort

@palisadoes  Yes, This can be fixed by :
- We can add a REDIS_URL field in the `.env` and `.sample.env` file
**OR**
- We can change our script so that it dose not display the `REDIS_PASSWORD`. For this, we can change the prompt to `REDIS URL` already exists.

palisadoes commented:
> > What can be done to fix? Why didn't it fail for the other prompts?
> > Please remember the registration message when configuring the email of last resort
> 
> @palisadoes Yes, This can be fixed by :
> 
>     * Adding a `REDIS_URL` field in the `.env` and `.sample.env` file
>       **OR**
> 
>     * Changing our script so that it dose not display the `REDIS_PASSWORD`. For this, we can change the prompt to `REDIS URL` already exists.

Don't display the password. We don't need another parameter to track.

palisadoes reviewed (changes requested):
- It detects an existing mongo URL, but you still have to enter it manually or else it fails
![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/3ba70315-e442-47ad-b45d-ebd2904e3d2c)

palisadoes reviewed (commented):
1. The sample database importation fails on the brand new system. Probably because the DB wasn't created. Please investigate.
![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/5e4af8d0-7474-442b-8342-81d20340518c)

palisadoes reviewed (changes requested):
1. The sample database importation fails on the brand new system. Probably because the DB wasn't created. Please investigate.
![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/5e4af8d0-7474-442b-8342-81d20340518c)

palisadoes commented:
1. We are almost there. Please take a look. I think I found an edge case
2. Also run tsdoc against the script to make sure we document all the functions for others

palisadoes commented:
> 1. It detects an existing mongo URL, but you still have to enter it manually or else it fails.
> 
>     2. This was done on a newly installed system. So we may want to make sure the database gets created after entering the URL
![image](https://private-user-images.githubusercontent.com/16875803/295652964-3ba70315-e442-47ad-b45d-ebd2904e3d2c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwODE0MDQsIm5iZiI6MTcwNTA4MTEwNCwicGF0aCI6Ii8xNjg3NTgwMy8yOTU2NTI5NjQtM2JhNzAzMTUtZTQ0Mi00N2FkLWI0NWQtZWJkMjkwNGUzZDJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEyVDE3MzgyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwMGJiNDBmNTM5YTU3MDk4ZWQ0OGY5ZjVhM2NkNGQ2N2YzODIzZGU5MGU1YjJmNGRkYzhlZDQ0MmJjMTAzMDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.J0h8yt7oS1RSpQHiEGFze0py5YclmPHAnEu_AO3FXmo)

This  still needs to be fixed

palisadoes reviewed (changes requested):
![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/91ea4f52-db54-4351-bd62-dcbbec21437e)

Please fix the DB importation and the mongodb url item too.

Otherwise this is ready to be merged

palisadoes commented:
1. Please fix the conflicting files. 
2. We merged a PR that required some `mongodb` checks for a GitHub action workflow to provision a demo Talawa-API cloud instance

palisadoes commented:
The database import problem seems to be caused by something else. Please focus on the remaining outstanding items
- https://github.com/PalisadoesFoundation/talawa-api/issues/1679#issuecomment-1890331968

Doraemon012 commented:
> The database import problem seems to be caused by something else. Please focus on the remaining outstanding items
> 
>     * [[BUG]: `import:sample-data` script ERR_REQUIRE_ESM #1679 (comment)](https://github.com/PalisadoesFoundation/talawa-api/issues/1679#issuecomment-1890331968)

- This issue is related to the use of `nanoid` (version 4 or higher) that does not support `CommonJS`. This is a known issue, and there's an issue open in the nanoid repository discussing it:

	- [Nanoid Issue #365](https://github.com/ai/nanoid/issues/365)
	
- Here's the relevant change log entry in the nanoid repository:

	- [Nanoid Change Log](https://github.com/ai/nanoid/blob/main/CHANGELOG.md#40)

- The problem is reflected in a merged Pull Request (PR) in the `talawa-api` repository, where `dependabot` automatically upgraded the nanoid version to `5.0.4`.
	
	- https://github.com/PalisadoesFoundation/talawa-api/pull/1561

- **To resolve the problem, we need to downgrade the version of nanoid to `3.3.7`.**

palisadoes commented on setup.ts:
```diff
@@ -95,38 +177,63 @@ async function askForRedisUrl(): Promise<{
       name: "email",
       message:
-        "Enter the email which you wish to assign as the Super Admin of last resort:",
+        "Please make sure to register with this email before logging in.\n Enter the email which you wish to assign as the Super Admin of last resort :",
```
This has the format of all the other questions. 

1. It is a statement, so it shouldn't start with a `?`
2. The bold text also makes it difficult to differentiate between other questions. All other non-questions are not bold
3. There should be spaces above and below the text to highlight this fact

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/543aa69e-564c-49f4-841a-221c97a31a3f)

palisadoes reviewed (changes requested):
See comments

palisadoes commented:
We are almost done.

Doraemon012 commented on setup.ts:
```diff
@@ -617,29 +809,47 @@ async function main(): Promise<void> {
-    });
-    if (shouldSetRedis) {
+    if (process.env.REDIS_HOST && process.env.REDIS_PORT) {
+      console.log(`\nRedis URL already exists`);
```
The `.env` file does not contain a parameter named `REDIS_URL`.

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/116624667/90fe97d1-9a3c-4e42-8cf3-e87793e79ae8)

palisadoes replied:
True. I could have been clearer. If a valid Redis configuration is detected, then all the Redis questions can be skipped. And if the valid URL is detected, the combined parameters to make the correct URL should be seen here. It makes it more intuitive.

Doraemon012 replied:
@palisadoes, as the `REDIS URL` contains the password too, the commit will fail a test. 
The URL is:
`url = redis://${password ? password + "@" : ""}${host}:${port}`

As the password is being displayed on the screen, the test fails.

palisadoes replied:
Can the password be replaced with `X` characters for display and to make the test pass?

palisadoes reviewed (changes requested):
See comment

Doraemon012 commented on setup.ts:
```diff
@@ -457,6 +634,12 @@ async function verifySmtpConnection(
+ * The function `configureSmtp` prompts the user to configure SMTP settings for sending emails through
+ * Talawa and saves the configuration in a .env file.
+ * @returns a Promise that resolves to void.
+ */
```
@palisadoes, please clarify what is the problem with the current logic.

palisadoes replied:
1. It asks whether you want to configure `SMTP` for emails.
2. You answer `No` and then it asks you for email information.

Doraemon012 replied:
I've created the functionality such that the user can configure either `mail` or `SMTP` (or neither) for mailing service.

palisadoes replied:
I see the issue. The code has an underlying flaw. We'll need to fix this later as a separate issue. 
Thanks

palisadoes commented:
@Doraemon012

How's progress?

palisadoes commented:
Let me know when you are ready for the final review.

palisadoes commented on setup.ts:
```diff
@@ -922,12 +907,16 @@ async function main(): Promise<void> {
   }
 
+  console.log(
+    "\n You can configure either SMTP or Mail for sending emails through Talawa.\n"
```
1. Should be `You can configure SMTP for sending emails through Talawa`
2. `SMTP` is a mail sending protocol

palisadoes replied:
Never mind. This can be resolved for the reasons stated previously

palisadoes commented:
To make this faster. I'm available on slack

palisadoes commented:
It looks like the only outstanding item is the Redis test

varshith257 commented:
Sorry, If I got wrong ...I understood the time taken to solve this issue by following all conversations of you and @Doraemon012 for this PR ...Just to make it successful merge can I follow up @Doraemon012 to setup it as quickly to close this issue.

Doraemon012 commented:
@palisadoes  I've made the requested changes. Please take a look.

palisadoes reviewed (changes requested):
1. The images folder must not be a part of this PR
2. Please merge with the latest upstream
3. This should be the last change.

palisadoes commented:
If not done so, make the loading of test data default to`No`

Doraemon012 commented:
@palisadoes I've merged my branch with the latest upstream and also set the default option for importing the `sample data` to `No`.
Please take a look.
